File: src/forager/ui/pages/game_grid.py
"""The scrollable card grid shown on the Library home page.

Owns the fixed-size portrait ``GameCard`` tiles, the adaptive column layout
(search-filtered, sorted), and keyboard/gamepad focus. Art tiles load
synchronously from local/cached sources here; network art is delivered later
by the worker thread and applied through :meth:`set_card_art`.
"""
from __future__ import annotations
from PySide6.QtCore import Qt, QEvent, Signal
from PySide6.QtWidgets import QWidget, QVBoxLayout, QLabel, QGridLayout, QScrollArea

from forager.core.game import Game
from forager.artwork import pipeline as art
from forager.library.metadata import filter_games, sort_key
from forager.ui import style
from forager.ui.theme import C
from forager.ui.widgets.game_card import GameCard

_GRID_MARGIN = 8
_GRID_MIN_GAP = 12
_GRID_V_GAP = 16


class GameGrid(QWidget):
    card_clicked = Signal(object)
    card_activated = Signal(object)
    layout_changed = Signal()

    def __init__(self, card_w: int, card_h: int, parent=None):
        super().__init__(parent)
        self._games: list[Game] = []
        self._cards: list[GameCard] = []
        self._card_index = 0
        self._card_w, self._card_h = card_w, card_h
        self._search_text = ""

        self.setStyleSheet("background: transparent;")
        self._build_ui()

    def _build_ui(self):
        v = QVBoxLayout(self)
        v.setContentsMargins(0, 0, 0, 0)
        v.setSpacing(0)

        self._empty_label = QLabel("No games found.")
        self._empty_label.setAlignment(Qt.AlignmentFlag.AlignCenter)
        style.label(self._empty_label, C.TEXT_DIM, size=14, padding="60px")
        v.addWidget(self._empty_label)

        self._scroll = QScrollArea()
        self._scroll.setWidgetResizable(True)
        self._scroll.setStyleSheet(
            "QScrollArea { background: transparent; border: none; }"
            "QScrollArea > QWidget > QWidget { background: transparent; }"
        )
        self._scroll.setHorizontalScrollBarPolicy(Qt.ScrollBarPolicy.ScrollBarAlwaysOff)
        self._scroll.viewport().installEventFilter(self)

        self._grid_host = QWidget()
        self._grid_host.setStyleSheet("background: transparent;")
        self._grid = QGridLayout(self._grid_host)
        self._grid.setContentsMargins(0, 0, 0, 0)
        self._grid.setSpacing(16)
        self._grid.setAlignment(Qt.AlignmentFlag.AlignTop | Qt.AlignmentFlag.AlignLeft)
        self._scroll.setWidget(self._grid_host)
        v.addWidget(self._scroll, stretch=1)

    # -- public API -----------------------------------------------------

    def set_games(self, games: list[Game]):
        self._games = games
        self._rebuild_cards()

    def set_search(self, text: str):
        self._search_text = text
        self._rebuild_cards()

    def set_card_size(self, card_w: int, card_h: int):
        self._card_w, self._card_h = card_w, card_h
        for card in self._cards:
            card.setFixedSize(card_w, card_h)
        self._relayout_cards()

    def cards(self) -> list[GameCard]:
        return self._cards

    def count(self) -> int:
        return len(self._cards)

    def needed_height(self) -> int:
        """Height the card grid needs right now (0 when empty)."""
        count = len(self._cards)
        if count == 0:
            return 0
        cols = self._grid.columnCount()
        if cols <= 0:
            return 0
        rows = (count + cols - 1) // cols
        return rows * self._card_h + (rows - 1) * _GRID_V_GAP

    def current_index(self) -> int:
        return self._card_index

    def card_at(self, index: int) -> GameCard | None:
        if 0 <= index < len(self._cards):
            return self._cards[index]
        return None

    def game_at(self, index: int) -> Game | None:
        card = self.card_at(index)
        return card.game if card is not None else None

    def column_count(self) -> int:
        return self._grid.columnCount()

    def viewport(self):
        return self._scroll.viewport()

    def focus_index(self, index: int):
        if not self._cards:
            return
        index = max(0, min(len(self._cards) - 1, index))
        for i, card in enumerate(self._cards):
            card.set_focused(i == index)
        self._card_index = index
        self._scroll.ensureWidgetVisible(self._cards[index], 40, 40)

    def set_card_art(self, game: Game, pix):
        for card in self._cards:
            if card.game == game:
                card.set_art(pix)
                return

    # -- internals ------------------------------------------------------

    def _filtered_games(self) -> list[Game]:
        return sorted(filter_games(self._games, self._search_text), key=sort_key)

    def _rebuild_cards(self):
        for card in self._cards:
            self._grid.removeWidget(card)
            card.deleteLater()
        self._cards.clear()
        self._card_index = 0

        for game in self._filtered_games():
            card = GameCard(game, card_w=self._card_w, card_h=self._card_h)
            card.clicked.connect(self.card_clicked)
            card.activated.connect(self.card_activated)
            self._cards.append(card)

        self._empty_label.setVisible(len(self._cards) == 0)
        self._scroll.setVisible(len(self._cards) > 0)
        self._relayout_cards()
        self._load_card_art()
        self.layout_changed.emit()

    def _load_card_art(self):
        for card in self._cards:
            card.set_art(art.load_grid(card.game, allow_network=False))

    def _relayout_cards(self):
        if not self._cards:
            return
        for i in reversed(range(self._grid.count())):
            widget = self._grid.itemAt(i).widget()
            if widget is not None:
                self._grid.removeWidget(widget)

        viewport_w = self._scroll.viewport().width()
        scrollbar = self._scroll.verticalScrollBar()
        if scrollbar is not None:
            viewport_w -= scrollbar.sizeHint().width()

        avail = max(1, viewport_w - 2 * _GRID_MARGIN)
        cols = max(1, (avail + _GRID_MIN_GAP) // (self._card_w + _GRID_MIN_GAP))
        cols = min(cols, len(self._cards))

        used = cols * self._card_w + (cols - 1) * _GRID_MIN_GAP
        remaining = avail - used
        if cols > 1 and remaining > 0:
            gap = _GRID_MIN_GAP + remaining // (cols - 1)
        else:
            gap = _GRID_MIN_GAP

        self._grid.setContentsMargins(_GRID_MARGIN, 0, _GRID_MARGIN, 0)
        self._grid.setHorizontalSpacing(gap)
        self._grid.setVerticalSpacing(_GRID_V_GAP)

        old_cols = getattr(self, "_layout_cols", 0)
        for col in range(max(old_cols, cols)):
            self._grid.setColumnStretch(col, 0)
        self._layout_cols = cols

        for i, card in enumerate(self._cards):
            self._grid.addWidget(card, i // cols, i % cols)

        self.layout_changed.emit()

    def eventFilter(self, obj, event):
        if obj is self._scroll.viewport() and event.type() == QEvent.Type.Resize:
            self._relayout_cards()
        return super().eventFilter(obj, event)

File: src/ui/pages/GameGrid.cpp
#include "ui/pages/GameGrid.h"
#include "ui/widgets/GameCard.h"
#include "ui/Theme.h"
#include "ui/Style.h"
#include "library/Metadata.h"
#include "artwork/Pipeline.h"
#include <QScrollBar>
#include <QEvent>
#include <algorithm>

using namespace theme;
using namespace theme::C;

static constexpr int GRID_MARGIN = 8;
static constexpr int GRID_MIN_GAP = 12;
static constexpr int GRID_V_GAP = 16;

GameGrid::GameGrid(int cardW, int cardH, QWidget* parent)
    : QWidget(parent), m_cardW(cardW), m_cardH(cardH)
{
    style::transparent(this);

    auto* v = style::vbox(this);

    m_emptyLabel = new QLabel("No games found.", this);
    m_emptyLabel->setAlignment(Qt::AlignCenter);
    style::label(m_emptyLabel, C::TEXT_DIM, 14);
    m_emptyLabel->setStyleSheet(m_emptyLabel->styleSheet() + "padding: 60px;");
    v->addWidget(m_emptyLabel);

    m_scroll = new QScrollArea(this);
    m_scroll->setWidgetResizable(true);
    m_scroll->setStyleSheet(style::scrollTransparentQss());
    m_scroll->setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);

    m_gridHost = new QWidget();
    style::transparent(m_gridHost);
    m_grid = new QGridLayout(m_gridHost);
    m_grid->setContentsMargins(0, 0, 0, 0);
    m_grid->setSpacing(GRID_V_GAP);
    m_grid->setAlignment(Qt::AlignTop | Qt::AlignLeft);
    m_scroll->setWidget(m_gridHost);
    v->addWidget(m_scroll, 1);

    m_scroll->viewport()->installEventFilter(this);
}

void GameGrid::setGames(const QList<Game>& games) {
    m_games = games;
    rebuildCards();
}

void GameGrid::setSearch(const QString& text) {
    m_searchText = text;
    rebuildCards();
}

void GameGrid::setCardSize(int cardW, int cardH) {
    m_cardW = cardW;
    m_cardH = cardH;
    for (auto* card : m_cards) card->setFixedSize(cardW, cardH);
    relayoutCards();
}

int GameGrid::count() const { return m_cards.size(); }

void GameGrid::setCardArt(const Game& game, const QPixmap& pix) {
    for (auto* card : m_cards) {
        if (card->game() == game) { card->setArt(pix); return; }
    }
}

void GameGrid::focusIndex(int index) {
    if (m_cards.isEmpty()) return;
    index = std::clamp(index, 0, static_cast<int>(m_cards.size()) - 1);
    for (int i = 0; i < m_cards.size(); ++i)
        m_cards[i]->setFocused(i == index);
    m_cardIndex = index;
    m_scroll->ensureWidgetVisible(m_cards[index], 40, 40);
}

void GameGrid::rebuildCards() {
    for (auto* card : m_cards) {
        m_grid->removeWidget(card);
        card->deleteLater();
    }
    m_cards.clear();
    m_cardIndex = 0;

    QList<Game> filtered = metadata::filterGames(m_games, m_searchText);
    std::sort(filtered.begin(), filtered.end(), [](const Game& a, const Game& b) {
        return metadata::sortKey(a) < metadata::sortKey(b);
    });

    for (const auto& game : filtered) {
        auto* card = new GameCard(game, m_cardW, m_cardH);
        connect(card, &GameCard::clicked, this, &GameGrid::cardClicked);
        connect(card, &GameCard::activated, this, &GameGrid::cardActivated);
        m_cards.append(card);
    }

    m_emptyLabel->setVisible(m_cards.isEmpty());
    m_scroll->setVisible(!m_cards.isEmpty());
    relayoutCards();
    loadCardArt();
    emit layoutChanged();
}

void GameGrid::loadCardArt() {
    for (auto* card : m_cards) {
        QPixmap pix = art::loadGrid(card->game(), false);
        if (!pix.isNull())
            card->setArt(pix);
    }
}

void GameGrid::relayoutCards() {
    if (m_cards.isEmpty()) return;

    while (m_grid->count() > 0) {
        auto* item = m_grid->takeAt(0);
        if (item->widget()) m_grid->removeWidget(item->widget());
        delete item;
    }

    int viewportW = m_scroll->viewport()->width();
    auto* sb = m_scroll->verticalScrollBar();
    if (sb) viewportW -= sb->sizeHint().width();

    int avail = std::max(1, viewportW - 2 * GRID_MARGIN);
    int cols = std::max(1, (avail + GRID_MIN_GAP) / (m_cardW + GRID_MIN_GAP));
    cols = std::min(cols, static_cast<int>(m_cards.size()));

    int used = cols * m_cardW + (cols - 1) * GRID_MIN_GAP;
    int remaining = avail - used;
    int gap = (cols > 1 && remaining > 0) ? GRID_MIN_GAP + remaining / (cols - 1) : GRID_MIN_GAP;

    m_grid->setContentsMargins(GRID_MARGIN, 0, GRID_MARGIN, 0);
    m_grid->setHorizontalSpacing(gap);
    m_grid->setVerticalSpacing(GRID_V_GAP);

    for (int i = 0; i < m_cards.size(); ++i)
        m_grid->addWidget(m_cards[i], i / cols, i % cols);

    emit layoutChanged();
}

bool GameGrid::eventFilter(QObject* obj, QEvent* event) {
    if (obj == m_scroll->viewport() && event->type() == QEvent::Resize)
        relayoutCards();
    return QWidget::eventFilter(obj, event);
}
